nav2_costmap_2d/src/segmentation_buffer.cpp

```diff
@@ -0,0 +1,264 @@
+  segmentation_list_.push_front(Segmentation());
+
+  // check whether the origin frame has been set explicitly
+  // or whether we should get it from the cloud
+  std::string origin_frame = sensor_frame_ == "" ? cloud.header.frame_id : sensor_frame_;
+
+  try
+  {
+    // given these segmentations come from sensors...
+    // we'll need to store the origin pt of the sensor
+    geometry_msgs::msg::PointStamped local_origin;
+    local_origin.header.stamp = cloud.header.stamp;
+    local_origin.header.frame_id = origin_frame;
+    local_origin.point.x = 0;
+    local_origin.point.y = 0;
+    local_origin.point.z = 0;
+    tf2_buffer_.transform(local_origin, global_origin, global_frame_, tf_tolerance_);
+    tf2::convert(global_origin.point, segmentation_list_.front().origin_);
+
+    sensor_msgs::msg::PointCloud2 global_frame_cloud;
+
+    // transform the point cloud
+    tf2_buffer_.transform(cloud, global_frame_cloud, global_frame_, tf_tolerance_);
+    global_frame_cloud.header.stamp = cloud.header.stamp;
+
+    // create a segmented pointcloud to store the original 3D data as well as the class and confidence
+    // of each point
+    sensor_msgs::msg::PointCloud2& segmentation_cloud = *(segmentation_list_.front().cloud_);
+    segmentation_cloud.height = global_frame_cloud.height;
+    segmentation_cloud.width = global_frame_cloud.width;
+    segmentation_cloud.fields = global_frame_cloud.fields;
+    segmentation_cloud.is_bigendian = global_frame_cloud.is_bigendian;
+    segmentation_cloud.point_step = global_frame_cloud.point_step;
+    segmentation_cloud.row_step = global_frame_cloud.row_step;
+    segmentation_cloud.is_dense = global_frame_cloud.is_dense;
+
+    unsigned int cloud_size = global_frame_cloud.height * global_frame_cloud.width;
+    sensor_msgs::PointCloud2Modifier modifier(segmentation_cloud);
+
+    // add the class and confidence fields to the pointcloud.
+    segmentation_cloud.point_step =
+      addPointField(segmentation_cloud, "class", 1, sensor_msgs::msg::PointField::INT8,
+                    segmentation_cloud.point_step);
+    segmentation_cloud.point_step =
+      addPointField(segmentation_cloud, "confidence", 1, sensor_msgs::msg::PointField::INT8,
+                    segmentation_cloud.point_step);
+    
+    // update the point step and get iterators
+    modifier.resize(cloud_size);
+    sensor_msgs::PointCloud2Iterator<uint8_t> iter_class_obs(segmentation_cloud, "class");
+    sensor_msgs::PointCloud2Iterator<uint8_t> iter_confidence_obs(segmentation_cloud, "confidence");
+    sensor_msgs::PointCloud2Iterator<float> iter_x_obs(segmentation_cloud, "x");
+    sensor_msgs::PointCloud2Iterator<float> iter_y_obs(segmentation_cloud, "y");
+    sensor_msgs::PointCloud2Iterator<float> iter_z_obs(segmentation_cloud, "z");
+
+    // get iterators for the original cloud
+    sensor_msgs::PointCloud2ConstIterator<float> iter_x_global(global_frame_cloud, "x");
```

Comment:
Same here with use of iterators, see the normal observation buffer as an example